### ISSUE TITLE
PDAL 1.9.0 breaks python filter on numpy (ubuntu build from src)

### ISSUE THREAD

**Issue by tomvantilburg** [PDAL/PDAL#2527]
1.8.0 src builds and runs fine
1.9.0 and 1.9.1 src build fine but throw on usage:
```
Original error was: /usr/local/lib/python2.7/dist-packages/numpy/core/_multiarray_umath.so: undefined symbol: PyExc_UserWarning

PDAL: ImportError('numpy.core.multiarray failed to import',)
```
When I use the git version (1.9-maintenance) the build breaks on:
```
[ 46%] Building CXX object plugins/python/filters/CMakeFiles/pdal_plugin_filter_python.dir/__/plang/Invocation.cpp.o
/var/data/git_repos/PDAL/plugins/python/plang/Invocation.cpp:40:10: fatal error: numpy/arrayobject.h: No such file or directory
 #include <numpy/arrayobject.h>
          ^~~~~~~~~~~~~~~~~~~~~
```
when doing the same on the master branch the build runs fine but running a python filter throws the import error again.

System info:
```
python: 2.7.15rc1
numpy: 1.16.3 (in /usr/local/lib/python2.7/dist-packages)
OS: Ubuntu 18.04
cmake: 3.10.2
```
I do have Python 3.6.7 installed as well.
Only using apt and pip

**Comment by abellgithub** [comment]
There are no changes WRT build for 1.9-maintenance from 1.9.1 or 1.9.  I think you'll need to determine what version of Python is being found by cmake (it should be listed as part of the build process) and also verify the library that's being linked by the libpdal_python_filter_python.<ext>.

Version 1.9 should prefer python 3.  You'll have to specify python 2 paths explicitly if you have both on your system.  I'm not sure that cmake 3.10 was explicitly tested, so there may be issues WRT the python find script for that version.  There was a lot of change going on in FindPython.cmake around that time.

**Comment by tomvantilburg** [comment]
Yes, I saw there were changes and also the other issues with it (#2444 and #2431) but cmake doesn't seem to pick up my python3 (python2.7 is the default version on my system).
It says:
```-- Found PythonInterp: /usr/bin/python (found version "2.7.15") ```

I just tested with manually configuring python 3.6 but get roughly same issue 
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/numpy/core/__init__.py", line 40, in <module>
    from . import multiarray
  File "/usr/local/lib/python3.6/dist-packages/numpy/core/multiarray.py", line 12, in <module>
    from . import overrides
  File "/usr/local/lib/python3.6/dist-packages/numpy/core/overrides.py", line 6, in <module>
    from numpy.core._multiarray_umath import (
ImportError: /usr/local/lib/python3.6/dist-packages/numpy/core/_multiarray_umath.cpython-36m-x86_64-linux-gnu.so: undefined symbol: PyExc_ImportError
```

Looking around on the internet it seems to be a numpy / cython issue that occurs more often where the library gets referenced before it is loaded but the solutions are unclear to me and it is hard to debug.

**Comment by abellgithub** [comment]
I would make make sure that the numpy library that's found matches the header files that are used to build PDAL.  There may be some information in the cmake cache file, but you may need to look at the python find script.  I built many configurations for release, so I know that it usually works out of the box, but there are many ways that python may be installed/configured and it's probably not possible for us to debug this remotely.

**Comment by tomvantilburg** [comment]
In the meantime PDAL picks up the correct python version (3.7) but still isn't loading it on time it seems.
I was able to work around it by adding `::dlopen("libpython3.7.so", RTLD_LAZY |RTLD_GLOBAL);` to the `loadPython()` routine in `./plugins/python/plang/Environment.cpp` 

Anything I can provide to make this a more robust solution?

**Comment by hobu** [comment]
if you export `PDAL_PYTHON_LIBRARY` to point to the hard path of your `libpython3.7.so`, it will load that directly.

**Comment by abellgithub** [comment]
You can also make a link called libPython.so -> libpython3.7.so

**Comment by tomvantilburg** [comment]
I couldn't get that to work: 
```
export PDAL_PYTHON_LIBRARY=/usr/lib/python3.7/config-3.7m-x86_64-linux-gnu/libpython3.7.so
make 
sudo make install
```

On a side node, there was this issue that dealt with seemingly the same problem? https://github.com/PDAL/PDAL/issues/1010

**Comment by abellgithub** [comment]
It's a runtime env. var.  Has nothing to do with build.

**Comment by tomvantilburg** [comment]
I managed to get an output of the content of `libname.data()` which is in my case `libPython` (so no extension):
```
if (libname.empty())
        libname = "libPython" + pdal::Utils::dynamicLibExtension;
    cout << libname.data();
    ::dlopen(libname.data(), RTLD_LAZY | RTLD_GLOBAL);
```
Apparently there is no dynamicLibExtension found here?
When I do `sudo ln -s ./libpython3.7.so ./libPython` it indeed works. 

Is this the way it is supposed to work?
The missing extension and camel-case of libPython look a bit off here.

**Comment by abellgithub** [comment]
Fixed by #2598
